javascript/upgrades/c8859f3725d4b070a877f8792214582d517c8a9b/toplevel_parent_xml_node.ql

```diff
@@ -1,18 +1,37 @@
+class Location extends @location {
+  string toString() { none() }
+
+  predicate startsAtLine(@file file, int line) { locations_default(this, file, line, _, _, _) }
+}
+
 class TopLevel extends @toplevel {
   string toString() { none() }
+
+  Location getLocation() { hasLocation(this, result) }
+
+  pragma[nomagic]
+  predicate startsAtLine(@file file, int line) { getLocation().startsAtLine(file, line) }
 }
 
 class XmlNode extends @xmllocatable {
   string toString() { none() }
+
+  Location getLocation() { xmllocations(this, result) }
+
+  pragma[nomagic]
+  predicate startsAtLine(@file file, int line) { getLocation().startsAtLine(file, line) }
 }
 
-// Based on previous implementation on HTMLNode.getCodeInAttribute and getInlineScript
+// Based on previous implementation on HTMLNode.getCodeInAttribute and getInlineScript,
+// with `startsAtLine` added for performance reasons.
 from
-  TopLevel top, XmlNode xml, @file f, @location l1, int sl1, int sc1, int el1, int ec1,
-  @location l2, int sl2, int sc2, int el2, int ec2
+  TopLevel top, XmlNode xml, @file f, Location l1, int sl1, int sc1, int el1, int ec1, Location l2,
+  int sl2, int sc2, int el2, int ec2
 where
-  xmllocations(xml, l1) and
-  hasLocation(top, l2) and
+  l1 = xml.getLocation() and
+  l2 = top.getLocation() and
+  xml.startsAtLine(f, sl1) and
+  top.startsAtLine(f, [sl1, sl1 + 1]) and
```

Comment:
```suggestion
  top.startsAtLine(f, sl2) and
```

Isn't this supposed to be `sl2`?  
And I don't understand the `+ 1`

Reply:
The point is to find `xml, top` pairs using fast join, which we can then test one at a time using the precise containment check (as the containment check does not give rise to a good join ordering).

For this purpose, we pick the pairs where `xml` and `top` start on the same line, or where `top` starts on the following line.